### ISSUE TITLE
Supporting longer VO names in multi-VO rucio?

### ISSUE THREAD

**Issue by sfayer** [rucio/rucio#4548]
Hi,

I would like to suggest adding a mapping of long VO names down to the three-letter VO tokens used by the database at the API level.

Motivation
----------
The motivation for this is to make the API easier for other projects to use. It also has the benefit of allowing users to put the full VO name in their local client config file (i.e. more user friendly too). One of my colleagues is working on a "multi-VO DIRAC" to "multi-VO Rucio" interface and this would make the API far easier to call from DIRAC.

Modification
------------
I've put together a proof-of-concept patch of how this could work:
https://github.com/sfayer/rucio/commit/5e2bd5ff389426bb713f066844136b52666fe2ad

It simply adds an optional VO mapping section to the rucio.cfg on the server and uses this to remap the VO name in the API functions. If a mapping isn't found (or the section doesn't exist), the vo field is passed through unchanged to keep this feature entirely optional and completely backwards compatible. This approach also avoids any database schema changes.

Would you please consider this as a possible addition to Rucio? I'd be happy to submit my code as a pull request or discuss other possible approaches to this.

Regards,
Simon

**Comment by bari12** [comment]
Hi Simon,

In principle something like this would work, but in some cases it is a bit more complicated, in others a bit easier than your proof-of-concept.

1. In many of the methods the `vo` parameter actually is derived from the used token (By looking up to which account, and therefore VO the token is associated to). In these cases, you do not have to do the mapping, it just has to be done first when the token is requested. Later on the `"short_vo_name"` is already part of this token handling via https://github.com/rucio/rucio/blob/e313d70d2cbb8f0c20b660afeb6f3c8117863a12/lib/rucio/web/rest/flaskapi/v1/common.py#L87
So in these cases (and I think these are the majoriy), it is enough to implement the mapping properly in the authentication layer. It does not have to be done later on in the REST/API layer.

2. In method where you manipulate VOs and to some extend also accounts, the mapping has to be part of the function, similar to how you do it. 

3. This should not be part of the `rucio.cfg` but part of the [config table](https://github.com/rucio/rucio/blob/master/lib/rucio/core/config.py) instead. It's pretty much the same interfaces, but this way it is easier to sync this within the infrastructure if you run many (hundreds?) of pods/VMs.

4. There might be one or the other special case I am missing, need to let this sink a little bit and re-iterate 😄 

For some other details you should get in touch with @ijjorama to coordinate these developments.

Cheers,
Martin

**Comment by sfayer** [comment]
Hi Martin,

Thanks for the suggestions... I (clearly) hadn't realised the VO was part of the auth token. I'll set-up a more complete development environment and have another look at the implementation. Please let me know if you think of anything else.

Regards,
Simon

**Comment by bari12** [comment]
It's not really part of the auth token, but it's associated with it. On each REST command, when checking if the auth-token is (still) valid, the method supplies the (short) vo name to the REST/API layer.

**Comment by sfayer** [comment]
Hi,

I've put together a new version of the patch which uses the database config table and only does the VO mapping at the main REST interface layer (and the few VO management commands where the user supplies a VO name directly as a parameter):
https://github.com/sfayer/rucio/commit/9abc5b0aa8de2a7b7ef898188e5bfa419c757ba2
Documentation:
https://github.com/sfayer/documentation/commit/50d8bb8018fe315d9ae1995e7bc8a6d8f0ece618

Do you have any further comments/ideas on this?

Regards,
Simon

**Comment by bari12** [comment]
I think this is good, but we should specify (and test this very well with lots of unit-tests) what characters are allowed in the long vo name; I don't have all code pathes in mind and I am not 100% sure if you could have spaces, special characterse, slashes etc.

**Comment by sfayer** [comment]
Hi,

The usual spec for long VO names is pretty much the same as plain DNS names (a-zA-Z0-9.-). When you say "we should specify", do you mean in the documentation or something else?

Regards,
Simon

**Comment by bziemons** [comment]
> The usual spec for long VO names is pretty much the same as plain DNS names (a-zA-Z0-9.-). When you say "we should specify", do you mean in the documentation or something else?

The documentation should mention that and the code should check & enforce that, imo.

**Comment by sfayer** [comment]
Sure, I've updated the documentation, added a check in the code to enforce the character set and added a thorough test for the accepted names:
https://github.com/sfayer/rucio/commit/a5cfd20be3e1f93dcba0a870d461f5d3c496c0b3
https://github.com/sfayer/documentation/commit/515d2cca059a7d9fad9337bc04c7aa456fcc1ed4

Would you be happy for me to send pull requests for these now?

Regards,
Simon

**Comment by bari12** [comment]
Hi @sfayer
Sorry for the delay. In general this looks ok and you can submit it for more comments from other, but in terms of testing I think we should essentially run the entire multi_vo test suite with these VO Name conversions, just to see if there is any functionality which does not work with it. I can't really see a reason why it wouldn't, but we should test the full spectrum of functions.

Thanks a lot for the effort!

**Comment by sfayer** [comment]
Thanks, I've opened a pull request with this patch for comments now.

It looks like there has been quite a big refactoring from pyweb -> flask since I started looking at this, so I've updated the code for that. I've also adjusted the full multi VO test suite to properly exercise the new code (which found a few extra edge cases that I'd missed up to now). I've also opened a pull request for the documentation for this feature.

Regards,
Simon
